src/python/pants/backend/experimental/python/lockfile.py

```diff
@@ -0,0 +1,113 @@
+from pants.engine.target import Targets
+from pants.util.strutil import pluralize
+
+logger = logging.getLogger(__name__)
+
+
+class PipToolsSubsystem(PythonToolBase):
+    options_scope = "pip-tools"
+    help = "Used to generate lockfiles for third-party Python dependencies."
+
+    default_version = "pip-tools==6.2.0"
+    default_main = ConsoleScript("pip-compile")
+
+    @classmethod
+    def register_options(cls, register):
+        super().register_options(register)
+        # TODO: How should users indicate where to save the lockfile to when we have per-tool
+        #  lockfiles and mmultiple user lockfiles?
+        register(
+            "--lockfile-dest",
+            type=str,
+            default="3rdparty/python/lockfile.txt",
+            help="The file path to be created.\n\nThis will overwrite any previous files.",
+        )
+
+    @property
+    def lockfile_dest(self) -> str:
+        return cast(str, self.options.lockfile_dest)
+
+
+class LockSubsystem(GoalSubsystem):
+    name = "lock"
+    help = "Generate a lockfile."
+
+
+class LockGoal(Goal):
+    subsystem_cls = LockSubsystem
+
+
+# TODO: What should the input to `./pants lock` be when generating user lockfiles? Currently, it's
+#  the 3rd-party requirements that get used as input. Likely it should instead be all 3rd party
+#  reqs used by the transitive closure? That would better mirror generate_lockfile.sh.
+@goal_rule
+async def generate_lockfile(
+    targets: Targets, pip_tools_subsystem: PipToolsSubsystem, workspace: Workspace
+) -> LockGoal:
+    reqs = PexRequirements.create_from_requirement_fields(
+        tgt[PythonRequirementsField] for tgt in targets if tgt.has_field(PythonRequirementsField)
+    )
+
+    input_requirements_get = Get(
+        Digest, CreateDigest([FileContent("requirements.in", "\n".join(reqs).encode())])
+    )
+    pip_compile_get = Get(
+        VenvPex,
+        PexRequest(
+            output_filename="pip_compile.pex",
+            internal_only=True,
+            requirements=PexRequirements(pip_tools_subsystem.all_requirements),
+            # TODO: Figure out which interpreter constraints to use...Among other reasons, this is
+            #  tricky because python_requirement_library targets don't have interpreter constraints!
+            interpreter_constraints=InterpreterConstraints(["CPython==3.9.*"]),
+            main=pip_tools_subsystem.main,
+        ),
+    )
+    input_requirements, pip_compile = await MultiGet(input_requirements_get, pip_compile_get)
+
+    dest = pip_tools_subsystem.lockfile_dest
+    result = await Get(
+        ProcessResult,
+        VenvPexProcess(
+            pip_compile,
+            description=(
+                f"Generate lockfile for {pluralize(len(reqs), 'requirements')}: {', '.join(reqs)}"
+            ),
+            argv=[
+                "requirements.in",
+                "--generate-hashes",
+                f"--output-file={dest}",
```

Comment:
Is this output format something that PEX can consume...? Should PEX be delegating to `pip_tools` in order to put the fetched stuff in its cache...?

Reply:
Oh, and speaking of caching: should this be configured to use a `named_cache` (until PEX is invoking it)?

Reply:
Interesting on caching. The --output-file is a literal requirements.txt file. 

I haven't investigated yet how pip-tools is doing the fetching and where it gets cached, but fwict it delegates to Pip. That makes sense to try to share that cache with Pex, and at a minimum to use a dedicated named_cache. I'll add a TODO in the next PR.